### PR TITLE
Feature/hocs 2907 compose up all

### PR DESCRIPTION
**Pull request by philbarton**
Improvements to allow 'docker-compose up' to work.

- Cleanup health checks
- Adjust health check timeouts
- Add memory options for workflow (can use 1.3G memory!)
- Use a mock server instead of clamav (saves 1.2G memory)
- Add a data migrate step lo load the hocs-data into the info schema
- Add a load elastic step to add the elastic indexes.
- Tested with 8G memory of Docker resource

All noted in the docker/README

### CODE REVIEW

File: docker/docker-compose.yml [hunk 1]
```diff
@@ -186,13 +194,16 @@ services:
...
+      start_period: 3m
+      timeout: 30s
+      retries: 3
     image: quay.io/ukhomeofficedigital/hocs-workflow:${WORKFLOW_TAG:-branch-main}
     ports:
       - 8091:8080
     networks:
       - hocs-network
     environment:
+      JAVA_OPTS: '-Xms1536M -Xmx2G'
```
Comment (zuzak) [inline docker/docker-compose.yml:206]: huh. is java not clever enough to know how much RAM to use? Do we need to do this for the other services too?

File: docker/docker-compose.yml [hunk 2]
```diff
@@ -64,16 +54,10 @@ services:
...
-        - CMD
-        - bash
-        - -c
-        - awslocal es list-domain-names
-          && awslocal s3 ls
-          && awslocal sqs list-queues
-      interval: 5s
-      timeout: 10s
-      start_period: 90s
+      test: "bash -c 'AWS_ACCESS_KEY_ID=fake AWS_SECRET_ACCESS_KEY=fake aws --endpoint-url=http://localhost:4572 s3 ls'"
```
Comment (zuzak) [inline docker/docker-compose.yml:57]: I'd define this as an array to make the line wrap quicker

File: docker/README.md [hunk 3]
```diff
@@ -65,6 +76,78 @@ $ docker-compose pull
 $ docker-compose up
 ```
 
+## Use of .env file
+The docker-compose file has variables for the image tags. This allows the use of a ``.env`` file.
+By default, ``branch-main`` will be used. The following is an example of a file overriding some services :
```
Comment (zuzak) [inline docker/README.md:81]:
```suggestion
By default, ``branch-main`` will be used. The following is an example of a file overriding some services:
```

It'd be better to default to `latest` as that's, well, the default behaviour.

File: .gitignore [hunk 4]
```diff
@@ -1,2 +1 @@
-.env
-.idea
\ No newline at end of file
+/docker/keycloak/hsperfdata_jboss/*
```
Comment (zuzak) [inline .gitignore:1]: You can't (easily) check in an empty folder using Git, but still:
```suggestion
/docker/keycloak/hsperfdata_jboss
```

File: docker/README.md [hunk 5]
```diff
@@ -65,6 +76,78 @@ $ docker-compose pull
...
+FRONTEND_TAG=branch-epic_HOCS-COMP
+CASEWORK_TAG=branch-epic_HOCS-COMP
+WORKFLOW_TAG=branch-epic_HOCS-COMP
+CASE_CREATOR_TAG=branch-epic_HOCS-COMP
+INFO_TAG=branch-epic_HOCS-COMP
+HOCS_DATA_TAG=branch-feature_HOCS-2907-compose-up-all
+HOCS_DATA_REPO=hocs-data
+HOCS_DATA_ELASTIC_REPO=hocs-data-elastic
+```
+The `HOCS_DATA_REPO` can be one of :
```
Comment (zuzak) [inline docker/README.md:92]: Colons don't generally have leading spaces. Don't make me dust off my Chicago Manual of Style.
```suggestion
The `HOCS_DATA_REPO` can be one of:
```

File: docker/README.md [hunk 6]
```diff
@@ -65,6 +76,78 @@ $ docker-compose pull
...
+* hocs-data
+* hocs-data-wcs
+
+The `HOCS_DATA_ELASTIC_REPO` can be one of :
```
Comment (zuzak) [inline docker/README.md:96]:
```suggestion
The `HOCS_DATA_ELASTIC_REPO` can be one of:
```

File: docker/README.md [hunk 7]
```diff
@@ -65,6 +76,78 @@ $ docker-compose pull
...
+* hocs-data-elastic
+* hocs-data-wcs-elastic
+
+## Data migration
+
+The info-service doesn&rsquo;t come with its schema, it relies on a data_migration step in the docker-compose file.
+
+The data_migration step loads the info schema using flyway scripts.
+
+To select hocs-data pr hocs-data-wcs set the variable ``HOCS_DATA_REPO`` in your ``.env`` file to either :
```
Comment (zuzak) [inline docker/README.md:106]:
```suggestion
To select hocs-data or hocs-data-wcs set the variable ``HOCS_DATA_REPO`` in your ``.env`` file to either:
```

File: .gitignore (hunk 4 above)
Comment (zuzak) [reply .gitignore:1]: As we're referencing `.env` in documentation and most developers will use it, it's okay to keep `.env` in the .gitignore.

File: docker/README.md [hunk 8]
```diff
@@ -65,6 +76,78 @@ $ docker-compose pull
...
+* hocs-data (default)
+* hocs-data-wcs
+
+The docker image for the data is stored in AWS ECR, to pull it you will need to set up
+an AWS profile and then run the following:
+```console
+$ $(aws ecr get-login --no-include-email --profile acp-ecr)
```
[the remark this answers is not in the document]
Comment (philbarton) [reply docker/README.md:113]: That gets the long login string and applies it.

File: docker/README.md [hunk 9]
```diff
@@ -65,6 +76,78 @@ $ docker-compose pull
...
+* hocs-data
+* hocs-data-wcs
+
+The `HOCS_DATA_ELASTIC_REPO` can be one of :
+* hocs-data-elastic
+* hocs-data-wcs-elastic
+
+## Data migration
+
+The info-service doesn&rsquo;t come with its schema, it relies on a data_migration step in the docker-compose file.
```
[the remark this answers is not in the document]
Comment (philbarton) [reply docker/README.md:102]: Changed

File: docker/README.md [hunk 10]
```diff
@@ -113,44 +196,20 @@ $ docker-compose up
...
 The frontend container doesn&rsquo;t handle authentication, as it is handled
 by a proxy before it hits the app. You&rsquo;ll need to set authentication
 headers yourself using (for example) a browser extension. Ask an existing
 developer for their copy. This might manifest itself as a browser timeout:
 check stderr before assuming something else is wrong.
+
+### Manually applying the Search indexes
+
+The search service needs elastic indexes applying. Without them, you will see continuous search errors.
+The search indexes are contained in either hocs-data or hocs-data-wcs. To apply the index, do th following:
```
[the remark this answers is not in the document]
Comment (philbarton) [reply docker/README.md:210]: corrected

File: docker/docker-compose.yml [hunk 11]
```diff
@@ -19,34 +19,24 @@ services:
...
-      - hocs-network
-
   clamav:
-    healthcheck:
-      # -f makes curl exit with 22 on error (it doesn't normally)
-      test: [ "CMD", "curl", "-f", "http://localhost:8080" ]
-    image: lokori/clamav-rest
-    environment:
-      - CLAMD_HOST=clamd
+    # Use a mockserver instead of bloated clamav
```
[the remark this answers is not in the document]
Comment (philbarton) [reply docker/docker-compose.yml:23]: Yeh, comes up really quickly and is only used when a document is added.

File: docker/README.md (hunk 3 above)
Comment (philbarton) [reply docker/README.md:81]: done

File: .gitignore (hunk 4 above)
Comment (philbarton) [reply .gitignore:1]: done

File: docker/README.md (hunk 6 above)
Comment (philbarton) [reply docker/README.md:96]: done

File: docker/README.md (hunk 7 above)
Comment (philbarton) [reply docker/README.md:106]: done

File: docker/README.md (hunk 5 above)
Comment (philbarton) [reply docker/README.md:92]: done

File: docker/docker-compose.yml [hunk 12]
```diff
@@ -127,9 +111,31 @@ services:
       localstack:
         condition: service_healthy
 
+  load_elastic:
+    image: 340268328991.dkr.ecr.eu-west-2.amazonaws.com/hocs/${HOCS_DATA_ELASTIC_REPO:-hocs-data-elastic}:${HOCS_DATA_TAG:-branch-main}
```
[the remark this answers is not in the document]
Comment (philbarton) [reply docker/docker-compose.yml:115]: I'd prefer to use branch-main as that is stable.
If I use latest I can get changes from feature branches, which could be anything!

[the remark this answers is not in the document]
Comment (LiamMacP) [reply docker/docker-compose.yml:115]: Latest and branch-main are the exact same, only on a push to main is the latest tag updated, at the same point branch-main would be updated to. 

I'd prefer latest too, as its more in line with general standards...

[the remark this answers is not in the document]
Comment (philbarton) [reply docker/docker-compose.yml:115]: Ok, I was basing my logic on that, without an explicit tag, docker defaults to latest. If we explicitly set tags everywhere then that's fine. 
I'll omit the tag, but it feels a bit ambiguous to me.

[the remark this answers is not in the document]
Comment (philbarton) [reply docker/docker-compose.yml:115]: Hhmm...The SHAs differ...between latest and branch-main..hey ho

[the remark this answers is not in the document]
Comment (LiamMacP) [reply docker/docker-compose.yml]: I'll look into why that's happened, we can always change back to `branch-main` if it causes issues.

File: docker/README.md [hunk 13]
```diff
@@ -1,19 +1,23 @@
 # docker-compose
 
 This directory contains configuration for Docker Compose, a quick way to get a
-local HOCS environment on your computer.
+local HOCS environment on your computer. 
+
+If this is the first time you run it, Compose will download the latest images; this could
+take fifteen minutes, depending on your connection. To relieve pressure on your machines resources,
```
Comment (zuzak) [inline docker/README.md:7]:
```suggestion
take fifteen minutes, depending on your connection. To relieve pressure on your machine&rsquo;s resources, 
```

File: docker/README.md [hunk 14]
```diff
@@ -65,6 +76,78 @@ $ docker-compose pull
 $ docker-compose up
 ```
 
+## Use of .env file
+The docker-compose file has variables for the image tags. This allows the use of a ``.env`` file.
+By default, ``branch-main`` will be used. The following is an example of a file overriding some services:
```
Comment (zuzak) [inline docker/README.md:81]:
```suggestion
By default, ``latest`` will be used. The following is an example of a file overriding some services:
```

File: docker/README.md [hunk 15]
```diff
@@ -65,6 +76,78 @@ $ docker-compose pull
...
+The docker image for the data is stored in AWS ECR, to pull it you will need to set up
+an AWS profile and then run the following:
+```console
+$ $(aws ecr get-login --no-include-email --profile acp-ecr)
+```
+
+To be able to pull the image you need to be in authorised to do so.
+This is in the ACP Hub in the Docker repos section.
+
+The AWS credentials are stored in the hub, under your "Connected Identities" section.
```
Comment (zuzak) [inline docker/README.md:119]:
```suggestion
AWS credentials are stored in the ACP hub, under your &ldquo;Connected Identities&rdquo; section.
```

File: docker/docker-compose.yml [hunk 16]
```diff
@@ -127,10 +111,32 @@ services:
...
+      FLYWAY_URL: jdbc:postgresql://postgres:5432/postgres
+      FLYWAY_USER: root
+      FLYWAY_PASSWORD: dev
+    networks:
+      - hocs-network
+    depends_on:
+      postgres:
+        condition: service_healthy
+
   frontend:
```
Comment (zuzak) [inline docker/docker-compose.yml:136]:
```suggestion
  frontend:
    healthcheck:
      test: [ "CMD", "wget", "-qo", "/dev/null",  "http://localhost:8080/actuator/health" ]
```

File: docker/README.md (hunk 15 above)
Comment (philbarton) [reply docker/README.md]: done

File: docker/README.md (hunk 14 above)
Comment (philbarton) [reply docker/README.md]: good spot, done

File: docker/README.md (hunk 13 above)
Comment (philbarton) [reply docker/README.md]: done

**Review by zuzak — APPROVED** [review]
Looks good. Slightly concerned this'll mean people not fully onboarded onto the dev team will be unable to run docker-compose, as it'll fail on the schema steps -- but I guess this was the case anyway

**Pull request merged**